### ISSUE TITLE
Pinocchio_The problem of inverse kinematics of robots

### ISSUE THREAD

**Issue by DawnQiu** [stack-of-tasks/pinocchio#1959]
Dear Pinocchio team,

I wanted to start off by expressing my gratitude for all the hard work you put into creating such a powerful and user-friendly tool. Pinocchio has been an invaluable asset to me, and I appreciate all the effort that went into its development.

However, I have encountered an issue while using Pinocchio for computing the inverse kinematics of my robot. Specifically, I am having trouble with the convergence of the results. 

To give some context, I am trying to use the end-effector frame of my robot to calculate its transformation matrix relative to the base frame. I have successfully imported my URDF and identified the end-effector frame index using the code `pinocchio::FrameIndex end_effector_id = model.getFrameId("EndEffector_FRAME")`. From there, I defined a desired pose in Cartesian space and set up an iterative algorithm to compute the necessary joint angles. 

However, despite adjusting the iteration parameters such as increasing the number of iterations and adjusting the time step, I am still unable to achieve convergence. Additionally, the resulting joint angle values exceed the physical limits of my robot.

I have attached my source code and output results for reference.

```
#include "pinocchio/parsers/sample-models.hpp"
#include "pinocchio/spatial/explog.hpp"
#include "pinocchio/algorithm/kinematics.hpp"
#include "pinocchio/algorithm/jacobian.hpp"
#include "pinocchio/algorithm/joint-configuration.hpp"
#include <pinocchio/parsers/urdf.hpp>
#include <pinocchio/algorithm/frames.hpp>
#include <pinocchio/spatial/se3.hpp>
#include <pinocchio/spatial/motion.hpp>
#include <pinocchio/spatial/explog.hpp>
#include <pinocchio/multibody/fwd.hpp>
#include <pinocchio/multibody/joint/joint-generic.hpp>


int main(int /* argc */, char ** /* argv */)
{
    // Load robot model from URDF file
    pinocchio::Model model;
    pinocchio::urdf::buildModel("/home/dawn/kortex-master/api_cpp/examples/GEN3Lite.urdf", model);
    pinocchio::Data data(model);

    // Get number of degrees of freedom of robot
    //const int nq = model.nq;

    // Get the end-effector placement
    pinocchio::FrameIndex end_effector_id = model.getFrameId("EndEffector_FRAME");

    // Set the Desired Pose
    const pinocchio::SE3 oMdes(Eigen::Matrix3d::Identity(), Eigen::Vector3d(0.55, 0., 0.4662));

    // Init the configuration
    Eigen::VectorXd q = pinocchio::neutral(model);


    const double eps  = 1e-4;
    const int IT_MAX  = 1000;
    const double DT   = 1e-1;
    const double damp = 1e-6;

    pinocchio::Data::Matrix6x J(6,model.nv);
    J.setZero();

    bool success = false;

    typedef Eigen::Matrix<double, 6, 1> Vector6d;
    Vector6d err;

    Eigen::VectorXd v(model.nv);

    for (int i=0;;i++)
    {
        pinocchio::forwardKinematics(model,data,q);
        pinocchio::updateFramePlacement(model,data,end_effector_id);
        const pinocchio::SE3 end_effector_transform = data.oMf[end_effector_id];
        const pinocchio::SE3 dMi = oMdes.actInv(end_effector_transform);
        err = pinocchio::log6(dMi).toVector();

        if(err.norm() < eps)
        {
            success = true;
            break;
        }
        if (i >= IT_MAX)
        {
            success = false;
            break;
        }

        // ComputeFrameJacobian
        pinocchio::computeFrameJacobian(model,data,q,end_effector_id,J);

        pinocchio::Data::Matrix6 JJt;
        JJt.noalias() = J * J.transpose();
        JJt.diagonal().array() += damp;
        v.noalias() = - J.transpose() * JJt.ldlt().solve(err);
        q = pinocchio::integrate(model,q,v*DT);
        
        if(!(i%10))
            std::cout << i << ": error = " << err.transpose() << std::endl;
    }

    if(success)
    {
        std::cout << "Convergence achieved!" << std::endl;
    }
    else
    {
        std::cout << "\nWarning: the iterative algorithm has not reached convergence to the desired precision" << std::endl;
    }

    std::cout << "\nresult: " << q.transpose() << std::endl;
    std::cout << "\nfinal error: " << err.transpose() << std::endl;
}
```
```
0: error =      -0.493  -0.0100005     0.48705 5.83072e-16           0           0
10: error =  -0.281649 -0.0461365   0.120834   0.126448   0.203841  -0.154534
20: error =  -0.280209   -0.15233  0.0319974 -0.0545198  0.0514053 -0.0724987
30: error =   -0.107308   -0.059766   0.0164938   0.0214273 -0.00372304   -0.032275
40: error =   -0.039057  -0.0201749  0.00536763  0.00717133 -0.00131633  -0.0108959
50: error =   -0.0698461  -0.00138607  -0.00141772   0.00213072 -0.000159297  -0.00350985
60: error =   -0.0262921 -0.000281498 -0.000585633  0.000731433 -5.31942e-05  -0.00125407
70: error =   -0.0324926    0.0019456   -0.0021677  0.000231455   2.9315e-05 -0.000448191
80: error =   -0.0669141   0.00541136  -0.00362962  6.92935e-05  0.000175136 -0.000143432
90: error =   -0.0254158   0.00206594  -0.00135524   2.4025e-05   6.3379e-05 -5.08754e-05
100: error =   -0.0698179   0.00581613  -0.00569061  1.19495e-05  0.000342873 -2.11805e-05
110: error =   -0.0262616   0.00219225  -0.00217706  4.20814e-06  0.000122011 -7.72361e-06
120: error =   -0.0466188   0.00394207  -0.00274578  2.43178e-06  0.000127918 -1.21585e-06
130: error =  -0.0245243  0.00207751 -0.00152676 1.10745e-06 5.82408e-05 -7.1742e-07
140: error =   -0.0520659   0.00441799  -0.00396572  1.90149e-06  0.000104858 -3.62002e-06
150: error =   -0.0241372   0.00204919  -0.00182805   6.7748e-07  4.64336e-05 -1.55379e-06
160: error =   -0.0280284    0.0023406  -0.00237312  1.91593e-06  0.000144282 -6.34552e-06
170: error =   -0.0273056   0.00230617  -0.00192574  1.07276e-06  8.17432e-05 -2.52449e-06
180: error =   -0.0242231   0.00205299  -0.00151676  7.71491e-07  5.56725e-05 -7.79051e-07
190: error =   -0.0536253    0.0045473  -0.00439802  2.29102e-06  0.000117943 -5.21944e-06
200: error =  -0.0337851  0.00286729  -0.0026992 1.17667e-06 6.77685e-05  -2.845e-06
210: error =   -0.0271465   0.00230481  -0.00184182  6.35519e-07  5.23062e-05 -9.31977e-07
220: error =   -0.0267772   0.00227432  -0.00178752  4.96143e-07  4.84366e-05 -7.64277e-07
230: error =   -0.0531302   0.00442106  -0.00427617  3.49233e-06   0.00030362 -1.17978e-05
240: error =   -0.0271943   0.00227916  -0.00214379  1.53848e-06  0.000120433 -4.76813e-06
250: error =   -0.063872  0.00532849 -0.00379418 2.47944e-06 0.000345435 7.13379e-06
260: error =  -0.0246268  0.00205934  -0.0014316 1.17856e-06 0.000123229 2.79914e-06
270: error =   -0.0388596   0.00328179  -0.00322461  2.46808e-06  0.000123021 -3.30867e-06
280: error =   -0.0241856   0.00204828  -0.00184591  1.02268e-06  6.11063e-05 -1.52755e-06
290: error =   -0.0292589   0.00248299  -0.00227211  9.73958e-07  5.92616e-05 -2.03458e-06
300: error =   -0.0245902    0.0020882  -0.00182721  6.00901e-07  4.54464e-05 -1.39301e-06
310: error =  -0.0361512  0.00298331 -0.00242798 1.40588e-06 0.000257321 -5.0003e-06
320: error =   -0.0274176   0.00229985  -0.00172025  1.09394e-06   0.00011655 -1.69785e-06
330: error =  -0.0322153  0.00272467  -0.0017843 1.14447e-06 8.98241e-05 2.81985e-07
340: error =   -0.0242134   0.00205306  -0.00147193  7.25779e-07  5.35383e-05 -1.98022e-07
350: error =   -0.0264608   0.00224631  -0.00162026  5.73012e-07  5.13509e-05 -1.78547e-07
360: error =  -0.0294731  0.00250208 -0.00171046 6.25699e-07 5.76218e-05 2.65927e-07
370: error =   -0.0257396   0.00218614  -0.00165943  5.88593e-07  4.65857e-05 -4.65963e-07
380: error =   -0.0241768   0.00205348  -0.00152388  4.98893e-07  4.37098e-05 -3.03431e-07
390: error =   -0.0269868   0.00228994  -0.00216723  9.72929e-07  5.57705e-05 -2.27898e-06
400: error =   -0.0369699   0.00313778  -0.00233708  8.22579e-07  7.45463e-05 -4.38033e-07
410: error =   -0.0513089   0.00434633  -0.00423822  2.57169e-06  0.000125795 -5.52528e-06
420: error =  -0.0252622  0.00214221 -0.00200423 9.11856e-07 5.63422e-05 -2.2398e-06
430: error =  -0.0333073    0.002823 -0.00190796 1.06149e-06 7.93262e-05 7.32456e-07
440: error =  -0.0245083   0.0019656 -0.00155225 2.57987e-07 0.000292748 3.16319e-06
450: error =  -0.0308139   0.0025796 -0.00209374  1.2859e-06 0.000141977 3.43877e-07
460: error =  -0.0276985  0.00234032 -0.00183228 9.78896e-07 7.96132e-05 -2.8495e-07
470: error =   -0.035745    0.002989 -0.00196803 1.47322e-06 0.000178926 5.15051e-06
480: error =  -0.0262818   0.0022169 -0.00159349 9.06968e-07 8.61709e-05 1.50998e-06
490: error =   -0.0440733   0.00373704  -0.00328668  1.58358e-06  9.64866e-05 -2.00785e-06
500: error =   -0.0330472   0.00280462  -0.00255949  1.05976e-06   6.6047e-05 -2.07385e-06
510: error =  -0.0426794  0.00361851  -0.0036372 1.90589e-06  9.6159e-05 -4.6499e-06
520: error =    -0.07312   0.0061862  -0.0042143 2.21096e-06 0.000200962  2.8959e-06
530: error =  -0.0273199  0.00231298 -0.00153515 1.00325e-06 7.18177e-05 1.15225e-06
540: error =  -0.0243503  0.00206574 -0.00141506 7.44457e-07 5.20148e-05 4.18725e-07
550: error =   -0.0242267   0.00205716  -0.00161921  5.77404e-07  4.55912e-05 -5.48788e-07
560: error =   -0.0267228   0.00226967  -0.00173738  4.86711e-07  4.84438e-05 -5.04679e-07
570: error =   -0.0528673   0.00444598  -0.00445353  3.45609e-06  0.000203781 -9.35307e-06
580: error =   -0.0256065   0.00216033   -0.0020887  1.36394e-06  8.29468e-05 -3.73109e-06
590: error =   -0.0258064   0.00218632  -0.00163734  9.23807e-07  6.26619e-05 -9.08312e-07
600: error =   -0.0276135   0.00234357   -0.0018711  6.76646e-07  5.54111e-05 -1.02678e-06
610: error =  -0.0356862  0.00293924 -0.00272289 1.70629e-06 0.000269119 2.07032e-07
620: error =   -0.0299841   0.00251643  -0.00227136  1.33342e-06  0.000124706 -1.04642e-06
630: error =   -0.0326153   0.00275862  -0.00269008  1.52484e-06  8.95287e-05 -2.69461e-06
640: error =  -0.0249955  0.00211943 -0.00192228  7.4706e-07 5.50404e-05 -1.6083e-06
650: error =  -0.0306377   0.0024725 -0.00219975 5.47747e-07 0.000332413 1.91045e-06
660: error =  -0.0260615  0.00217082 -0.00165726  1.0418e-06 0.000144837 4.48439e-07
670: error =  -0.0452989  0.00383061 -0.00360641  2.1832e-06 0.000125949 -3.2976e-06
680: error =   -0.0407915   0.00345802  -0.00270371   1.1214e-06  9.32009e-05 -9.34966e-07
690: error =   -0.0567556   0.00480619  -0.00471472  2.96625e-06  0.000143314 -6.22691e-06
700: error =   -0.0273728   0.00232058  -0.00220031  1.01214e-06  6.28806e-05 -2.53733e-06
710: error =   -0.0298998   0.00253662  -0.00184017  8.56832e-07  6.41082e-05 -2.35375e-07
720: error =  -0.0245069  0.00208054 -0.00146725 6.27669e-07 4.79725e-05 -3.8104e-08
730: error =   -0.0279118   0.00237041  -0.00197797  7.27349e-07   5.1273e-05 -1.19251e-06
740: error =  -0.0275267  0.00233704 -0.00166736  6.0617e-07 5.33137e-05 -1.3685e-08
750: error =  -0.0410465  0.00346772  -0.0021067 1.41215e-06  0.00012542 3.61685e-06
760: error =   -0.0391367   0.00331657  -0.00279711  1.64243e-06  9.29155e-05 -9.56157e-07
770: error =   -0.0356146   0.00302225  -0.00268091  1.19837e-06  7.23409e-05 -1.86033e-06
780: error =   -0.0245488   0.00208451  -0.00176556  5.66411e-07  4.61882e-05 -1.05528e-06
790: error =   -0.0250676   0.00212893  -0.00162357  5.91103e-07   4.5928e-05 -4.94427e-07
800: error =   -0.0457288   0.00388011  -0.00369407  1.71167e-06   9.3674e-05 -3.97032e-06
810: error =   -0.0283948   0.00241069  -0.00209215  6.74697e-07  5.48477e-05 -1.60076e-06
820: error =   -0.0269126   0.00228575  -0.00190386  5.16456e-07  4.88617e-05 -1.17707e-06
830: error =   -0.0249823   0.00212217  -0.00175131  5.25369e-07  4.38711e-05 -1.03974e-06
840: error =   -0.0262521   0.00223011  -0.00182286  5.51594e-07  4.57233e-05 -1.02833e-06
850: error =  -0.0609976   0.0051575 -0.00327814 1.73969e-06 0.000173941 4.17331e-06
860: error =  -0.0241823  0.00204662 -0.00129395 8.62512e-07   6.444e-05 1.53938e-06
870: error =   -0.0302191    0.0025598   -0.0024259  1.50001e-06  7.56848e-05 -2.41101e-06
880: error =   -0.0248392   0.00201194  -0.00192485  1.19502e-06  0.000256354 -6.04328e-06
890: error =   -0.0358908   0.00301602  -0.00224474  1.51943e-06  0.000142919 -2.10007e-06
900: error =   -0.0259404   0.00219278  -0.00166564   8.7028e-07  7.28922e-05 -1.03761e-06
910: error =   -0.0622967   0.00527467  -0.00520663  3.19359e-06  0.000157958 -7.12425e-06
920: error =   -0.0242975   0.00205872  -0.00204825  1.08232e-06  5.88353e-05 -2.74691e-06
930: error =   -0.0241535    0.0020501  -0.00173197  6.74483e-07  4.78535e-05 -1.32398e-06
940: error =   -0.0310633   0.00263229  -0.00264858  1.48034e-06  7.41931e-05 -3.67962e-06
950: error =   -0.0569686   0.00482879  -0.00481659  2.68885e-06  0.000131373 -6.40592e-06
960: error =   -0.0266532     0.002261  -0.00218587  9.00871e-07   5.7257e-05 -2.57486e-06
970: error =   -0.0314408   0.00266966   -0.0023435  8.89692e-07  5.86769e-05 -1.96623e-06
980: error =  -0.0241839  0.00205417 -0.00172568 4.98766e-07 4.32033e-05  -1.147e-06
990: error =   -0.0808512      0.00684  -0.00699465  4.34226e-06  0.000217189 -1.10219e-05

Warning: the iterative algorithm has not reached convergence to the desired precision

result: -0.0856276   -1.64197   -0.29861    39.2699   -1.34329   -39.1843

final error:    -0.029882   0.00252969  -0.00262287  1.54569e-06  7.76168e-05 -4.12558e-06
```

I am wondering if you have any insights into what may be causing this issue. Could there be an error in my code, or is there something else that I am overlooking? Any guidance or suggestions would be greatly appreciated.

Thank you again for all your hard work and dedication to creating such a valuable tool.

Best regards!

**Comment by stephane-caron** [comment]
Be careful that your error vector here is expressed in the target frame `d` ($log_6({}_a M_b)$ is a vector in $a$):

```c++
err = pinocchio::log6(dMi).toVector();
```

But your Jacobian is in the body frame `i`:

```c++
pinocchio::computeFrameJacobian(model,data,q,end_effector_id,J);
```

Error and Jacobian should be in the same frame, e.g. in the body (end-effector) frame [as in this note](https://scaron.info/robotics/jacobian-of-a-kinematic-task-and-derivatives-on-manifolds.html#pose-task-residual).

I checked in the docs in this inconsistency is [already in the example](https://gepettoweb.laas.fr/doc/stack-of-tasks/pinocchio/master/doxygen-html/md_doc_b-examples_d-inverse-kinematics.html#autotoc_md44). Perhaps the author of the example can chime in? Or shall we correct it to put everything in the body frame?

**Comment by jcarpent** [comment]
You can also change the `dt` to do smaller steps.

**Comment by DawnQiu** [comment]
Thank you for getting back to me. I think I understand what you're saying now. The issue is that the error coordinate system is different from the Jacobian matrix coordinate system.

The error reference coordinate system is based on the desired end-effector pose coordinate system, while the Jacobian matrix's reference coordinate system is based on the base coordinate system. However, I'm not sure how to solve this problem. Do you have any suggestions on how I can align the err reference coordinate system with the base coordinate system?

Thank you again for your help and I look forward to hearing from you.

**Comment by stephane-caron** [comment]
@DawnQiu You can check out #1963 for a change set to apply to the example to get everything in the local frame (NB: not reviewed yet).

**Comment by jcarpent** [comment]
Fix in #1963
